fs/cifs/cifsacl.c

```diff
@@ -1670,3 +1680,1056 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+{
+	switch (smbacetype) {
+	case ACCESS_ALLOWED_ACE_TYPE:
+		*nfs_ace_type_out = ACE4_ACCESS_ALLOWED_ACE_TYPE;
+		break;
+	case ACCESS_DENIED_ACE_TYPE:
+		*nfs_ace_type_out = ACCESS_DENIED_ACE_TYPE;
+		break;
+	default:
+		cifs_dbg(VFS, "ACE contains unsupported ace type 0x%04x\n",
+			 smbacetype);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/*
+ * Convert SID into NFS4 ID and type. Try to map BUILTIN / SPECIAL sids
+ * directly to NFS4 special type where possible (to avoid upcall to winbindd).
+ *
+ * Since the existing idmap key implementation does support return of both
+ * users and groups in one call, we first try to retrieve group (because this
+ * is in real life much more likely). If retrieving as group fails, we retry
+ * as user.
+ */
+static int
+convert_smb_sid_to_nfs_who_special(struct cifs_sid *psid,
+				   u32 *iflag,
+				   u32 *who_id,
+				   u32 *flags)
+{
+	/*
+	 * Check for direct mapping of owner@, group@, and everyone@
+	 */
+	if (psid->num_subauth > 3) {
+		return 0;
+	}
+
+	if (compare_sids(psid, &sid_everyone) == 0) {
+		*iflag = ACEI4_SPECIAL_WHO;
+		*who_id = ACE4_SPECIAL_EVERYONE;
+		return 1;
+	}
+
+	if (compare_sids(psid, &sid_creator_owner) == 0) {
+		*iflag = ACEI4_SPECIAL_WHO;
+		*who_id = ACE4_SPECIAL_OWNER;
+		return 1;
+	}
+
+	if (compare_sids(psid, &sid_creator_group) == 0) {
+		*iflag = ACEI4_SPECIAL_WHO;
+		*who_id = ACE4_SPECIAL_GROUP;
+		*flags |= ACE4_IDENTIFIER_GROUP;
+		return 1;
+	}
+
+	/*
+	 * SID communicating Unix mode can be safely skipped since we will
+	 * get permissions info from other ACL entries
+	 */
+	if (compare_sids(psid, &sid_unix_NFS_mode) == 0)
+		return -EAGAIN;
+
+	/*
+	 * SID may directly encode a Unix uid or gid. We can short-circuit
+	 * this check if number of SID subauths does not match what's expected
+	 * in this case (which will result in upcall to winbind).
+	 */
+	if (psid->num_subauth == 2) {
+		// Possibly Samba encoding of unix id
+		if (psid->sub_auth[0] == sid_unix_groups.sub_auth[0]) {
+			*flags |= ACE4_IDENTIFIER_GROUP;
+			*who_id = le32_to_cpu(psid->sub_auth[1]);
+			return 1;
+		} else if (psid->sub_auth[0] == sid_unix_users.sub_auth[0]) {
+			*who_id = le32_to_cpu(psid->sub_auth[1]);
+			return 1;
+		}
+	} else if (psid->num_subauth == 3) {
+		if (psid->sub_auth[1] == sid_unix_NFS_groups.sub_auth[1]) {
+			*flags |= ACE4_IDENTIFIER_GROUP;
+			*who_id = le32_to_cpu(psid->sub_auth[2]);
+			return 1;
+		} else if (psid->sub_auth[1] == sid_unix_users.sub_auth[1]) {
+			*who_id = le32_to_cpu(psid->sub_auth[2]);
+		}
+	}
+
+	return 0;
+}
+
+static int
+convert_smb_sid_to_nfs_who(struct cifs_sid *psid, u32 *iflag, u32 *who_id, u32 *flags)
+{
+	char *sidstr;
+	const struct cred *saved_cred;
+	struct key *sidkey;
+	uint sidtype = SIDGROUP;
+	int rc;
+
+	if (unlikely(psid->num_subauth > SID_MAX_SUB_AUTHORITIES)) {
+		cifs_dbg(FYI, "%s: %u subauthorities is too many!\n",
+			 __func__, psid->num_subauth);
+		return -EIO;
```

Comment:
Good point. I don't like that comment either (copy-pasted check from other area). -EINVAL is probably better choice. Since this eventually gets back to xattr handler ERANGE might confuse user.

I'll probably need to go through and think over / justify errors used.